### PR TITLE
Add ability to open storage session outside the pipeline

### PR DESCRIPTION
**Pull request by danielmarbach**
This back ports the functionality of v8 to resolve a completable storage session from a child container and then open it for consumption outside the pipeline. 

This behavior is required for the transactional session package.

### CODE REVIEW

File: src/NServiceBus.AcceptanceTests/Core/Persistence/When_a_persistence_does_not_provide_synchronized_storage_session.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@
     using Unicast.Subscriptions;
     using Unicast.Subscriptions.MessageDrivenSubscriptions;
 
-    public class When_a_persistence_does_not_provide_ISynchronizationContext : NServiceBusAcceptanceTest
+    public class When_a_persistence_does_not_provide_synchronized_storage_session : NServiceBusAcceptanceTest
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Core/Persistence/When_a_persistence_does_not_provide_synchronized_storage_session.cs:13]: Maybe you can port this also to master?

File: src/NServiceBus.AcceptanceTests/Core/Persistence/When_a_persistence_provides_synchronized_session.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+namespace NServiceBus.AcceptanceTests.Core.Persistence
+{
+    using System.Threading.Tasks;
+    using AcceptanceTesting;
+    using EndpointTemplates;
+    using NServiceBus.Persistence;
+    using NUnit.Framework;
+
+    public class When_a_persistence_provides_synchronized_session : NServiceBusAcceptanceTest
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Core/Persistence/When_a_persistence_provides_synchronized_session.cs:9]: Maybe you can port this also to master?

File: src/NServiceBus.AcceptanceTests/Core/Reliability/SynchronizedStorage/When_opening_storage_session_outside_pipeline.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    using System.Threading.Tasks;
+    using AcceptanceTesting;
+    using EndpointTemplates;
+    using Extensibility;
+    using Features;
+    using NUnit.Framework;
+    using ObjectBuilder;
+    using Persistence;
+
+    public class When_opening_storage_session_outside_pipeline : NServiceBusAcceptanceTest
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Core/Reliability/SynchronizedStorage/When_opening_storage_session_outside_pipeline.cs:13]: Maybe you can port this also to master?

File: src/NServiceBus.Core/Reliability/SynchronizedStorage/CompletableSynchronizedStorageSessionAdapter.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        public async Task<bool> TryOpen(TransportTransaction transportTransaction, ContextBag context)
+        {
+            AdaptedSession = await synchronizedStorageAdapter.TryAdapt(transportTransaction, context).ConfigureAwait(false);
+
+            return AdaptedSession != null;
+        }
+
+        public async Task Open(ContextBag contextBag) => AdaptedSession = await synchronizedStorage.OpenSession(contextBag).ConfigureAwait(false);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Reliability/SynchronizedStorage/CompletableSynchronizedStorageSessionAdapter.cs:38]: I think we are inconsistent with calling in context and contextBag. This is also not entirely aligned on master. Maybe you can double check that

**Review by timbussmann — APPROVED** [review]
looks ok to me

File: src/NServiceBus.AcceptanceTests/Core/Reliability/SynchronizedStorage/When_opening_storage_session_outside_pipeline.cs (hunk 3 above)
Comment (danielmarbach) [reply src/NServiceBus.AcceptanceTests/Core/Reliability/SynchronizedStorage/When_opening_storage_session_outside_pipeline.cs:12]: https://github.com/Particular/NServiceBus/pull/6530 done

File: src/NServiceBus.Core/Reliability/SynchronizedStorage/CompletableSynchronizedStorageSessionAdapter.cs (hunk 4 above)
Comment (danielmarbach) [reply src/NServiceBus.Core/Reliability/SynchronizedStorage/CompletableSynchronizedStorageSessionAdapter.cs]: Done

**Review by tmasternak — APPROVED** [review]

**Pull request merged**